src/Data/List/Properties.agda

```diff
@@ -740,6 +714,50 @@ map-concatMap f g xs = begin
   concatMap (map f ∘′ g) xs
     ∎
 
+------------------------------------------------------------------------
+-- catMaybes
+
+catMaybes-concatMap : catMaybes {A = A} ≗ concatMap fromMaybe
+catMaybes-concatMap [] = refl
+catMaybes-concatMap (mx ∷ xs)
+  with IH ← catMaybes-concatMap xs
+  with mx
+... | just x  = cong (x ∷_) IH
+... | nothing = IH
+
+length-catMaybes : ∀ xs → length (catMaybes {A = A} xs) ≤ length xs
+length-catMaybes [] = ≤-refl
+length-catMaybes (mx ∷ xs)
+  with IH ← length-catMaybes xs
+  with mx
+... | just _  = s≤s IH
+... | nothing = m≤n⇒m≤1+n IH
+
+catMaybes-++ : (xs ys : List (Maybe A)) →
+               catMaybes (xs ++ ys) ≡ catMaybes xs ++ catMaybes ys
+catMaybes-++ []        _  = refl
+catMaybes-++ (mx ∷ xs) ys
+  with IH ← catMaybes-++ xs ys
+  with mx
+... | just x  = cong (x ∷_) IH
+... | nothing = IH
```

Comment:
Notwithstanding your discussion with @JacquesCarette , I fail to see how these proofs improve on the original three-liners of L126-L149. What am I not grasping here? Esp. given your noted/notable aversion to `with`-based programming...

Reply:
I used irrefutable `with` for reusability: the induction hypothesis is now only written once and shared across the branches of the case analysis. (the new feature for [left-hand side lets via the `using` keyword](https://github.com/agda/agda/pull/7078) would be even better, but this will have to wait until the next release)

Re. my `with`-scepticism, it mostly refers to basic definitions like `mapMaybe` and not so frequently when you prove properties about them. Especially when a use of `with` is unnecessary, that is you do not really need *dependent* pattern matching; here you actual need the dependency. All at the same time considering the reduction behaviour which might now favour the use of `with`, as mentioned in [@gallais' comment](https://github.com/agda/agda-stdlib/issues/1937#issuecomment-1483389966) in the original discussion about this.

 **However**, just to contradict my own argument above, when one gets into the business of proving meta-properties about these properties (e.g. you view proofs of membership [constructively as indices](https://omelkonian.github.io/formal-prelude/Prelude.Lists.Membership.html) --- ironically I also needed that for [permutations](https://omelkonian.github.io/formal-prelude/Prelude.Lists.PermutationsMeta.html) --- also c.f. notes about "meta-properties" in my [thesis](https://omelkonian.github.io/data/phd-thesis.pdf)), there is no distinction between "base" definitions and "properties". Having said that, the standard library does not include any of these to my knowledge, so we can just ignore these ultra-rare cases for the sake of practicality.

Therefore, I still think the current form sits nicely between the eliminator-style approach requested by Jacques and the original fully-direct non-reusing pattern matching.

Reply:
I'm sympathetic to the argument based on sharing, but where it falls down for me is in the simple-minded LOC here... and some kind of trade-off in cognitive load that follows from that... sharing a recursive call both affords, but also *enforces*, a particular understanding of the call structure which is quite distinct from the one-clause-one-call-structure style. You might argue it's 'richer', but you also pay a higher price for it!

My instinct for `stdlib` goes towards a more 'middlebrow' approach... with a correspondingly lower bar to entry for newcomers?

Reply:
As for 'meta' properties, I find increasingly the need for proofs of 'higher-dimensional' lifting lemmas (eg in [`Data.List.Relation.Binary.BagAndSetEquality`](https://github.com/agda/agda-stdlib/blob/c5255d0c28ca39b3233f4ade7c21a0a96b7a5b69/src/Data/List/Relation/Binary/BagAndSetEquality.agda#L579-L592)...

Reply:
There are actually already occurrences of the `with ih ← ...` pattern already in the codebase + one could say that this is more intuitive to beginners than simultaneous deep pattern matching maybe?

Reply:
Indeed! ... I'm responsible for them... ;-)